crates/ruff_linter/src/rules/flake8_return/rules/function.rs

```diff
@@ -473,76 +498,73 @@ fn implicit_return(checker: &mut Checker, stmt: &Stmt) {
-                    content.push_str(checker.stylist().line_ending().as_str());
-                    content.push_str(indent);
-                    content.push_str("return None");
-                    diagnostic.set_fix(Fix::unsafe_edit(Edit::insertion(
-                        content,
-                        end_of_last_statement(stmt, checker.locator()),
-                    )));
+                if checker.settings.preview.is_disabled() {
+                    add_return_none(checker, stmt, stmt.range());
                 }
-                checker.diagnostics.push(diagnostic);
+                true
             }
         }
         Stmt::Match(ast::StmtMatch { cases, .. }) => {
+            let mut result = false;
             for case in cases {
                 if let Some(last_stmt) = case.body.last() {
-                    implicit_return(checker, last_stmt);
+                    if has_implicit_return(checker, last_stmt) {
+                        result = true;
+                        if checker.settings.preview.is_enabled() {
+                            return result;
+                        }
+                    }
                 }
             }
+            result
         }
         Stmt::With(ast::StmtWith { body, .. }) => {
             if let Some(last_stmt) = body.last() {
-                implicit_return(checker, last_stmt);
+                if has_implicit_return(checker, last_stmt) {
+                    return true;
+                }
             }
+            false
         }
-        Stmt::Return(_) | Stmt::Raise(_) | Stmt::Try(_) => {}
+        Stmt::Return(_) | Stmt::Raise(_) | Stmt::Try(_) => false,
         Stmt::Expr(ast::StmtExpr { value, .. })
             if matches!(
                 value.as_ref(),
                 Expr::Call(ast::ExprCall { func, ..  })
                     if is_noreturn_func(func, checker.semantic())
-            ) => {}
+            ) =>
+        {
+            false
+        }
         _ => {
-            let mut diagnostic = Diagnostic::new(ImplicitReturn, stmt.range());
-            if let Some(indent) = indentation(checker.locator(), stmt) {
-                let mut content = String::new();
-                content.push_str(checker.stylist().line_ending().as_str());
-                content.push_str(indent);
-                content.push_str("return None");
-                diagnostic.set_fix(Fix::unsafe_edit(Edit::insertion(
-                    content,
-                    end_of_last_statement(stmt, checker.locator()),
-                )));
+            if checker.settings.preview.is_disabled() {
+                add_return_none(checker, stmt, stmt.range());
             }
-            checker.diagnostics.push(diagnostic);
+            true
         }
     }
 }
 
+/// RET503
+fn implicit_return(checker: &mut Checker, function_def: &ast::StmtFunctionDef, stmt: &Stmt) {
+    if has_implicit_return(checker, stmt) && checker.settings.preview.is_enabled() {
```

Comment:
I don't have a great solution but I do find it a bit surprising that `has_implicit_return` adds diagnostics in the non preview mode. 

I'm somewhat inclined to duplicate the logic between preview and non-preview mode. That should also make it much easier when stabilizing the new behavior (and it's not that much code that needs copying). Unless you see a way of how we can avoid the side-effects of `has_implicit_return` in stable (e.g. could it return a `Vec` with all the statements for which diagnostics should be added?)

Reply:
I don't mind going ahead with the PR as is but I'm interested to hear what you think of duplicating the logic OR maybe you have another idea to make this a bit clearer

Reply:
I use the `Vec` solution to remove the side effects. 
The code is nicer.
Thank you for suggesting it.

Reply:
Thanks. Will merge now :)

Reply:
You are welcome